### PR TITLE
[FIX] sale_project: fix `_compute_tasks_ids` for non-billable project

### PR DESCRIPTION
**Pull request by pivi-odoo**
## Issue
There is a traceback when confirming/opening some SOs form view about indexing a boolean.

## Steps to reproduce
- Install Sales, Project
- Create a non-billable project
- Create a service product that on SO confirmation creates a Task, and the project is our newly created non-billable project
- Install Timesheets
- Create a new SO with the service
- Upon confirmation of the SO, traceback.

## Cause
Following https://github.com/odoo/odoo/pull/136255, which was making the implicit assumption that tasks that have a `sale_line_id` set *should* have a `sale_order_id` associated also, but apparently it is not the case always, you can have tasks associated with an SOL, but have the SO removed from the compute in `sale_timesheet`:
```py
@api.depends('allow_billable')
def _compute_sale_order_id(self):
    billable_tasks = self.filtered('allow_billable')
    super(ProjectTask, billable_tasks)._compute_sale_order_id()
    (self - billable_tasks).sale_order_id = False
```
which removes the `sale_order_id` of non-billable tasks, but not the `sale_line_id`.
In that inconsistent state, the `_read_group` when grouping by has entries where the `sale_order_id` is False, and indexing on it will throw the traceback.

## Fix
Since the behaviour of `allow_billable` is not consistent between the module `sale_project` and `sale_timesheet` (in the former you can have an SO of non-billable project, but in the latter you can't), we are just handling this use-case in the
`_compute_tasks_ids` to fetch the SO from the `sale_line_id`, as it's the least invasive solution for a stable patch.

## Affected versions
16.0 up to master

## Reference
opw-3531041

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pivi-odoo** [comment]
Alternative solutions would be:
- Write `sale_line_id` to `False` in the `_compute_sale_order_id`.
- Correct the domain of the read_group to not take into account tasks that have a `sale_line_id`, but no `sale_order_id` set on it, but this would be a change of behaviour.

File: addons/sale_project/models/sale_order.py [hunk 1]
```diff
@@ -60,10 +60,17 @@ def _compute_tasks_ids(self):
...
-        }
+        so_to_tasks_and_count = {}
+        for group in tasks_per_so:
+            if group['sale_order_id']:
+                so_to_tasks_and_count[group['sale_order_id'][0]] = {'task_ids': group['ids'], 'count': group['sale_order_id_count']}
+            else:
+                # tasks that have no sale_order_id need to be associated with the SO from their sale_line_id
+                for task in self.env['project.task'].browse(group['ids']):
+                    so_to_tasks_item = so_to_tasks_and_count.setdefault(task.sale_line_id.order_id.id, {'task_ids': [], 'count': 0})
+                    so_to_tasks_item['task_ids'] += task.ids
```
Comment (xavierbol) [inline addons/sale_project/models/sale_order.py:71]:
```suggestion
                    so_to_tasks_item['task_ids'].append(task.id)
```
? 😇

**Review by xavierbol — APPROVED** [review]
Thanks for your work, I left one comment and after that it will look good to me. 🙂 

robodoo delegate=pivi-odoo

Feel free to merge it once it is done. 🙂

**Comment by pivi-odoo** [comment]
@robodoo r+

**Pull request closed without merging**